app/actions/remote/thread.ts

```diff
@@ -194,3 +199,152 @@ export const updateThreadFollowing = async (serverUrl: string, teamId: string, t
         return {error};
     }
 };
+
+enum Direction {
+    Up,
+    Down,
+}
+
+async function fetchBatchThreads(
+    serverUrl: string,
+    teamId: string,
+    options: FetchThreadsOptions,
+    pages?: number,
+): Promise<{error: unknown; data?: Thread[]}> {
+    const operator = DatabaseManager.serverDatabases[serverUrl]?.operator;
+
+    if (!operator) {
+        return {error: `${serverUrl} database not found`};
+    }
+
+    let client: Client;
+    try {
+        client = NetworkManager.getClient(serverUrl);
+    } catch (error) {
+        return {error};
+    }
+
+    // if we start from the begging of time (since = 0) we need to fetch threads from newest to oldest (Direction.Down)
+    // if there is another point in time, we need to fetch threads from oldest to newest (Direction.Up)
+    let direction = Direction.Up;
+    if (options.since === 0) {
+        direction = Direction.Down;
+    }
+
+    const currentUser = await getCurrentUser(operator.database);
+    if (!currentUser) {
+        return {error: 'currentUser not found'};
+    }
+
+    const {config} = await getCommonSystemValues(operator.database);
+    const data: Thread[] = [];
+
+    const fetchThreadsFunc = async (opts: FetchThreadsOptions) => {
+        let page = 0;
+        const {before, after, perPage = General.CRT_CHUNK_SIZE, deleted, unread, since} = opts;
+
+        page += 1;
+        const {threads} = await client.getThreads(currentUser.id, teamId, before, after, perPage, deleted, unread, since, false, config.Version);
+        if (threads.length) {
+            // Mark all fetched threads as following
+            for (const thread of threads) {
+                thread.is_following = true;
+            }
+
+            data.push(...threads);
+
+            if (threads.length === perPage) {
```

Comment:
I heard at some point that it is possible to receive less than perPage elements and still have elements to fetch. I don't remember the context where this happened, but just wanted to make sure you are positive this does not happen here.

Also, a really corner case that we should probably not worry about is the case we apply a "perPage" higher than the server "perPage". In that case, we always will return a lower amount than perPage.

Reply:
I see that, at least for threads, the default page size is 30 only if we don't send a `perPage` param.
I see no reason why it shouldn't return as many as we ask each time.

@larkox do you have any more info on that?

Reply:
Now that I think about it, maybe it is only for searches (which use stuff like ElasticSearch, and so on), so probably we shouldn't worry about it. Marking as resolved.